ios/Trifle/Sources/TrifleCertificate.swift

```diff
@@ -0,0 +1,97 @@
+//
+//  TrifleCertificate.swift
+//  Trifle
+//
+
+import Foundation
+import Wire
+
+public class TrifleCertificate : Codable {
+    private let proto: Certificate
+    
+    public init(data: Data) throws {
+        self.proto = try ProtoDecoder().decode(Certificate.self, from: data)
+    }
+    
+    internal func getCertificate() -> Certificate {
+        return proto
+    }
+    
+    /**
+     Verify that the provided certificate matches what we expected.
+     It matches the CSR that we have and the root cert is what
+     we expect.
+
+     - parameters: certificateRequest request used to generate this certificate
+     - parameters: certificateChain - list of certificates between this cert and
+        the root certificate.
+     - parameters: rootCertificate - certificate to use as root of chain.
+        Defaults to the root certificate bundled with Trifle.
+     
+     - returns: true if validated, false otherwise
+     */
+    public func verify(
+        certificateRequest: MobileCertificateRequest? = nil,
+        intermediateTrifleChain: Array<TrifleCertificate>,
+        rootTrifleCertificate: TrifleCertificate? = nil
+    ) throws -> Bool {
+        // TODO: validate PK in MobileCertificateRequest against certificate
+        
+        let intermediateChain = intermediateTrifleChain.map({ trifleCert in
+            return trifleCert.getCertificate()
+        })
+        
+        return try self.proto.verify(certificateRequest: certificateRequest,
+                                     intermediateChain: intermediateChain,
+                                     rootCertificate: rootTrifleCertificate?.getCertificate())
+    }
+}
+
+
+extension Certificate {
+    /**
+     Verify that the provided certificate matches what we expected.
+     It matches the CSR that we have and the root cert is what
+     we expect.
+
+     - parameters: certificateRequest request used to generate this certificate
+     - parameters: certificateChain - list of certificates between this cert and
+        the root certificate.
+     - parameters: rootCertificate - certificate to use as root of chain.
+        Defaults to the root certificate bundled with Trifle.
+     
+     - returns: true if validated, false otherwise
+     */
+    internal func verify(
+        certificateRequest: MobileCertificateRequest? = nil,
+        intermediateChain: Array<Certificate>,
+        rootCertificate: Certificate? = nil
+    ) throws -> Bool {
+        
+        // TODO: validate PK in MobileCertificateRequest against certificate
+        
+        let chain : Array<Certificate>
+        if (rootCertificate != nil) {
+            chain = [self] + intermediateChain + [rootCertificate!]
+        } else {
+            chain = [self] + intermediateChain
+        }
+        
+        let result = X509TrustManager.evaluate(chain)
```

Comment:
I think whats bugging me is that while this is on the protobuf abstraction, there is some product logic tied into this as seen in this line where we have an x509 implicity. 

Trifle Certificate -> Certificate Proto -> certificate bytes field (maybe x509).

can we break this up even further? i.e. if certificateProto.version == 0 -> verifyX509(...)

otherwise, lets keep the implementation in 1 place under TrifleCertificate#verify

Reply:
I actually don't have a problem with the way this is structured i think. I can of course add in a case statement for version 0 being x509 but right now we only have one version and adding that abstraction doesnt really add anything. Unless Im misunderstanding it, you're saying in Certificate#verify add in a condition if v0 then Certificate#verifyX509.

If you feel strongly about this, I can add the condition and the extra method. If not, I am comfortable with the current structure.

Reply:
TrifleCertificate#verify conditionally branches to a particular Certificate#verify[X509|other encoding] based on versioning. Ideally, I would not want to tie Certificate#verify to one particular encoding strategy if we're thinking of abstractions.

Reply:
Ive captured this in a JIRA (725).